client/app.jsx

```diff
@@ -0,0 +1,114 @@
+const React = require('react'),
+      ReactDOM = require('react-dom'),
+      i18n = require('i18next-client'),
+      io = require('socket.io-client'),
+      _ = require('lodash'),
+      config = require('./config.jsx'),
+      Route = require('react-router-dom').Route,
+      HashRouter = require('react-router-dom').HashRouter,
+      Switch = require('react-router-dom').Switch,
+      TopBar = require('./topbar.jsx'),
+      Ting = require('./ting.jsx'),
+      Settings = require('./settings.jsx');
+
+const App = React.createClass({
+    _socket: null,
+    getInitialState() {
+        return {
+            username: null,
+            people: [],
+            currentMessageId: null
+        };
+    },
+    componentWillMount() {
+        const URL = window.location.hostname + ':' + config.port;
+        this._socket = io.connect(URL, {secure: config.websocket.secure});
+    },
+    updateUsername(username) {
+        this.setState({username});
+    },
+    updateCurrentMessageId(messageid) {
+        this.setState({currentMessageId: messageid});
+    },
+    onLoginIntention(username, people) {
+        this.setState({username, people});
+    },
+    listenOnLoginResponse(onError, onSuccess, onLogin, username) {
+        this._socket.on('login-response', ({success, people, error}) => {
+            if (!success) {
+                onError(error);
+            }
+            else {
+                onSuccess();
+
+                var peopleList = _.chain(people)
+                    .value();
+
+                onLogin(this.state.username, peopleList);
+            }
+        });
+    },
+    listenOnMessage(onMessage) {
+        this._socket.on('message', (data) => {
+            onMessage(data);
+        });
+    },
+    listenOnPart(onPart, deleteTypingMessage) {
+        this._socket.on('part', (username) => {
+            onPart(username);
+            deleteTypingMessage(username);
+        });
+    },
+    listenOnJoin(onJoin) {
+        this._socket.on('join', (username) => {
+            onJoin(username);
+        });
+    },
+    listenOnStartTypingResponse(onStartTypingResponse) {
+        this._socket.on('start-typing-response', (messageid) => {
+            this.setState({currentMessageId: messageid});
+            onStartTypingResponse(messageid);
+        });
+    },
+    listenOnUpdateTypingMessages(onUpdateTypingMessages) {
+        this._socket.on('update-typing-messages', (messagesTyping) => {
+            onUpdateTypingMessages(messagesTyping);
+        });
+    },
```

Comment:
This is not sufficient. Consider the following "race condition" case:

1. Alice opens up the chat
2. Bob opens up the chat
3. Alice starts typing a message reading "Hello, world" but without submitting it
4. Bob is informed that Alice has typed "Hello, world"
5. Bob opens up the Settings page
6. All of Bob's local "typing" state is destroyed by the router unmounting the Ting component and subsequently the History component.
7. Bob closes the Settings page and goes back to the chat window
8. Alice now submits her message by pressing enter
9. On Bob's side, the message never appears. He gets this error: `Uncaught TypeError: Cannot set property 'message_content' of undefined`

By the way, the above bug also exists without your PR – if Alice is typing a message, stops typing, and then Bob enters the Ting website and then Alice submits a message. So it's not something introduced by your code, but only made more obvious by the way we handle lost state.

If we want this to work properly, we need to proceed in one of two directions:

1. Either maintain the chat state (such as messages being typed, message history, and so on) regardless of whether the Settings component is open or not.
2. Only maintain the chat state when the chat window is open. Whenever the chat window closes, no state is maintained. When the chat window is reopened, any state required is re-requested from the server.

I think following the latter approach will be better because our code changes will be shorter and at the same time we will be able to reuse this logic when we switch to using multiple channels.

Therefore, I recommend we make sure that lack of state can be recovered from. But this is a matter for a next PR, no changes needed here. Perhaps @VitSalis would find this an interesting challenge to tackle.

Reply:
Will be resolved in a future PR as detailed in #213.